From 84c1204ed3733b23fbe7c8061d7014cba9797a8b Mon Sep 17 00:00:00 2001
From: Jacek Caban <jacek@codeweavers.com>
Date: Fri, 22 May 2026 00:51:27 +0200
Subject: [PATCH 10/48] ntdll: Allow calling NtContinue in syscall callbacks.

CW-Bug-Id: #27308

(cherry picked from commit 6a083d04cbd3bdc88d32c80a99432bb0a50e2654)
---
 dlls/ntdll/unix/signal_arm64.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/dlls/ntdll/unix/signal_arm64.c b/dlls/ntdll/unix/signal_arm64.c
index c7ed488f702..c18e6507f6c 100644
--- a/dlls/ntdll/unix/signal_arm64.c
+++ b/dlls/ntdll/unix/signal_arm64.c
@@ -345,13 +345,14 @@ NTSTATUS signal_set_full_context( CONTEXT *context )
 {
     struct syscall_frame *frame = get_syscall_frame();
     struct arm64_thread_data *arm64_data = arm64_thread_data();
+    CHPE_V2_CPU_AREA_INFO *cpu_area = NtCurrentTeb()->ChpeV2CpuAreaInfo;
     NTSTATUS status;
 
-    if (arm64_data->suspend_pending)
+    if (arm64_data->suspend_pending && !cpu_area->InSyscallCallback && !cpu_area->InSimulation)
     {
         sigset_t old_set;
         pthread_sigmask( SIG_BLOCK, &server_block_set, &old_set );
-        *NtCurrentTeb()->ChpeV2CpuAreaInfo->SuspendDoorbell = 0;
+        *cpu_area->SuspendDoorbell = 0;
         arm64_data->suspend_pending = FALSE;
         wait_suspend( context );
         status = NtSetContextThread( GetCurrentThread(), context );
-- 
2.55.0

